# Metrics App

[![Python](https://img.shields.io/badge/python-3.12.1-blue.svg)](https://www.python.org/downloads/release/python-3121/)
[![Docker](https://img.shields.io/badge/docker-available-blue)](https://www.docker.com/)

## Description

**Metrics App** is a Python and Streamlit-based application for performing genetic analyses, focused on evaluating gene and exon coverage from BAM/CRAM files. The app enables the selection and analysis of gene panels, individual genes, and entire exomes using BED files and read depth data generated by SAMtools. It also supports custom analyses, parallelization, and interactive reports for comprehensive genetic data analysis.

## Project Structure

```
metrics_app
├── app
│   ├── Home.py
│   ├── app_pages
│   │   ├── about.py
│   │   ├── gene_panel_creator.py
│   │   ├── logout.py
│   │   ├── query.py
│   │   ├── results.py
│   │   └── settings.py
│   └── components
│       ├── analysis.py
│       ├── bam_cram.py
│       ├── forms.py
│       ├── gene_panel_creator.py
│       ├── genome.py
│       ├── metrics.py
│       ├── plot.py
│       ├── s3.py
│       ├── samtools.py
│       ├── session_state.py
│       ├── streamlit_page_config.py
│       └── universal_bed.py
├── data
│   ├── depth
│   │   ├── exome
│   │   ├── gene_panel
│   │   └── single_gene
│   ├── img
│   │   ├── thumbnail_image001.png
│   │   ├── unilabs-1.svg
│   │   └── unilabs_logo.png
│   ├── mapped
│   └── regions
│       ├── gene_panels
│       └── genome_exons
├── environment.yml
├── Dockerfile
└── README.md
```

## Features

- **Gene Panel Analysis**: Enables the creation and analysis of custom gene panels.
- **Single Gene Analysis**: Provides detailed analysis of specific genes, evaluating exon coverage.
- **Exome Analysis**: Supports analysis of all genes and exons in an exome.
- **Interactive Reports**: Generates interactive coverage plots and reports using Plotly.
- **S3 Integration**: (In Development) Integration with S3 for storing and retrieving BAM/CRAM files.
- **Parallel Processing**: (In Development) Uses parallel processing to optimize the analysis of large datasets.

## Prerequisites

- **Python 3.12.1**
- **Docker**
- **Conda** (optional) for managing dependencies.

### Python Dependencies

The dependencies for the project are listed in the `environment.yml` file:

```yaml
name: metrics_app_env
channels:
  - defaults
  - conda-forge
dependencies:
  - python=3.12.1
  - numpy=1.26.3
  - pandas=2.1.4
  - plotly=5.24.1
  - boto3=1.34.150
  - weasyprint=62.3
  - streamlit=1.38.0
  - openpyxl=3.1.2
  - cairo
  - pango
  - jpeg
  - gdk-pixbuf
  - pip=23.3.1
  - pip:
    - numpy==1.26.3
    - pandas==2.1.4
    - plotly==5.24.1
    - boto3==1.34.150
    - weasyprint==62.3
    - streamlit==1.38.0
    - openpyxl==3.1.2

```

## Installation and Setup

### Using Docker

1. **Clone the repository**:

```bash
git clone https://github.com/pfcv99/metrics_app.git
cd metrics_app
```

2. **Build the Docker image**:

```bash
docker build -t unilabs_app .
```

3. **Run the Docker container**:

```bash
sudo docker run -it -p 8501:8501 unilabs_app bash
```

4. **Start the Streamlit application inside the container**:

```bash
streamlit run app/Home.py
```

### Using Conda

1. **Create the Conda environment**:

```bash
conda env create -f environment.yml && conda clean -a
conda activate metrics_app
```

2. **Run the application**:

```bash
streamlit run app/Home.py
```

## Application Pages

- **Home**: Main page of the application providing an overview of the features.
- **Query**: Allows selection and configuration of gene, gene panel, and exome analyses.
- **Results**: Displays the results and interactive reports from the analyses.
- **Gene Panel Creator**: A tool for creating custom gene panels.
- **Settings**: Application settings, including S3 integration (in development).
- **Logout**: User management, including login and logout functionalities.

## Contribution

If you want to contribute to the development of this application, please follow these steps:

1. Fork the repository.
2. Create a new branch for your changes: `git checkout -b my-branch`.
3. Commit your changes: `git commit -m 'My changes'`.
4. Push the changes to your branch: `git push origin my-branch`.
5. Open a pull request.

## License

This project is licensed under the terms of the [MIT License](LICENSE).

## Contact

Developer: **Pedro Venâncio**  
Email: [pedrofcvenancio@ua.pt](mailto:pedrofcvenancio@ua.pt)