# Metrics App

[![Python](https://img.shields.io/badge/python-3.12.1-blue.svg)](https://www.python.org/downloads/release/python-3121/)
[![Docker](https://img.shields.io/badge/docker-available-blue)](https://www.docker.com/)

## Description

**Metrics App** is a bioinformatics application developed in Python and Streamlit to perform genetic analyses, focusing on the evaluation of gene and exon Depth of Coverage and Breadth of Coverage from BAM/CRAM files. The app supports the analysis of gene panels, individual genes, and entire exomes using BED files and depth data generated by SAMtools. It includes interactive reports for comprehensive genetic data visualization and plans to incorporate parallel processing to optimize performance on large datasets.

## Project Structure

```
metrics_app
├── Dockerfile
├── README.md
├── __init__.py
├── environment.yml
├── app
│   ├── Home.py
│   ├── __init__.py
│   ├── app_pages
│   │   ├── __init__.py
│   │   ├── about.py
│   │   ├── gene_panel_creator.py
│   │   ├── logout.py
│   │   ├── query.py
│   │   ├── results.py
│   │   └── settings.py
│   └── components
│       ├── __init__.py
│       ├── analysis.py
│       ├── bam_cram.py
│       ├── forms.py
│       ├── gene_panel_creator.py
│       ├── genome.py
│       ├── metrics.py
│       ├── plot.py
│       ├── s3.py
│       ├── samtools.py
│       ├── session_state.py
│       ├── streamlit_page_config.py
│       └── universal_bed.py
└── data
   ├── depth
   │   ├── exome
   │   ├── gene_panel
   │   └── single_gene
   ├── img
   │   ├── thumbnail_image001.png
   │   ├── unilabs-1.svg
   │   └── unilabs_logo.png
   ├── mapped
   └── regions
       ├── gene_panels
       └── universal_bed
           ├── GRCh37_hg19
           └── GRCh38_hg38
```

## Features

- **Gene Panel Analysis**: Analyze Depth and Breadth of Coverage across custom gene panels.
- **Single Gene Analysis**: Provides in-depth analysis of individual genes and exons, evaluating both Depth and Breadth of Coverage.
- **Exome Analysis**: Supports analysis of entire exomes, computing metrics across all genes and exons.
- **Interactive Reports**: Generates interactive reports and visualizations using Streamlit and Plotly, facilitating exploration of genetic data.
- **S3 Integration**: (In Development) Enables retrieval of BAM/CRAM files from S3 storage.
- **Parallel Processing**: (In Development) Optimizes analysis of large datasets through parallel processing.

## Prerequisites

- **Python 3.12.1**
- **Docker**
- **Conda** (optional) for managing dependencies.

### Python Dependencies

The project dependencies are listed in the `environment.yml` file:

```yaml
name: metrics_app_env
channels:
  - defaults
  - conda-forge
dependencies:
  - python=3.12.1
  - numpy=1.26.3
  - pandas=2.1.4
  - plotly=5.24.1
  - boto3=1.34.150
  - weasyprint=62.3
  - streamlit=1.38.0
  - openpyxl=3.1.2
  - cairo
  - pango
  - jpeg
  - gdk-pixbuf
  - pip=23.3.1
  - pip:
    - samtools==1.19.2
    - git
    - numpy==1.26.3
    - pandas==2.1.4
    - plotly==5.24.1
    - boto3==1.34.150
    - weasyprint==62.3
    - streamlit==1.38.0
    - openpyxl==3.1.2
```

## Installation and Setup

### Using Docker

1. **Clone the repository**:

```bash
git clone https://github.com/pfcv99/metrics_app.git
cd metrics_app
```

2. **Build the Docker image**:

```bash
docker build -t unilabs_app .
```

3. **Run the Docker container**:

```bash
sudo docker run -it -p 8501:8501 unilabs_app bash
```

4. **Start the Streamlit application inside the container**:

```bash
streamlit run app/Home.py
```

The application will be accessible at `http://localhost:8501`.

### Using Conda

1. **Clone the repository**:

```bash
git clone https://github.com/pfcv99/metrics_app.git
cd metrics_app
```

2. **Create the Conda environment**:

```bash
conda env create -f environment.yml
conda activate metrics_app_env
```

3. **Run the application**:

```bash
streamlit run app/Home.py
```

## Required Files

To ensure the application runs correctly, the following files (or equivalent) must be added to the respective directories:

- `data/regions/universal_bed/GRCh37_hg19/*.BED`
- `data/regions/universal_bed/GRCh38_hg38/*.BED`
- `data/regions/gene_panels/*.xlsx`
- `data/mapped/*.bam`

These files are necessary for the genetic analysis processes, and while the exact filenames may differ, they must be present in the specified folders.

## Application Pages

- **Home**: Main landing page.
- **Query**: Configure and execute genetic analyses for individual genes, gene panels, and exomes.
- **Results**: Displays interactive reports and analysis results.
- **Gene Panel Creator**: A tool for creating custom gene panels.
- **Settings**: Application settings, including S3 integration (in development).
- **Logout**: User management, including login and logout functionalities (in development).

## Contribution

To contribute to this project:

1. Fork the repository.
2. Create a new branch: `git checkout -b my-branch`.
3. Commit your changes: `git commit -m 'My changes'`.
4. Push to your branch: `git push origin my-branch`.
5. Open a pull request.

## License

This project is licensed under the terms of the [MIT License](LICENSE).

## Contact

Developer: **Pedro Venâncio, MSc in Clinical Bioinformatics, University of Aveiro**  
Email: [pedrofcvenancio@ua.pt](mailto:pedrofcvenancio@ua.pt)